### ISSUE TITLE
Please export track names in midi file

### ISSUE THREAD

**Issue by rualark** [paulrosen/abcjs#327]
Currently midi files look like this:
```
MFile 1 5 480
MTrk
0 Tempo 333333
0 Meta TrkEnd
TrkEnd
MTrk
0 PrCh ch=1 p=0
1920 On ch=1 n=65 v=105
3825 On ch=1 n=65 v=0
3825 Meta TrkEnd
TrkEnd
MTrk
0 PrCh ch=1 p=0
0 Meta TrkEnd
TrkEnd
MTrk
0 PrCh ch=1 p=0
0 Meta TrkEnd
TrkEnd
MTrk
0 PrCh ch=1 p=0
0 Meta TrkEnd
TrkEnd
```

**Comment by paulrosen** [comment]
Ok.

**Comment by paulrosen** [comment]
As long as you are examining the MIDI file output, I have a couple questions I'd like your opinion on:
1) I'm using an "ON" event with a volume of 0 to turn a note off instead of using an OFF event. That actually makes the file smaller because I don't need to repeat the command byte, but it only saves 1 byte per note. Is there any downside to doing that?

2) I have a separate track for some meta data. I don't remember why it was set up like that. The example you showed above doesn't contain a title, but if you used the `T:` field it would show up there, too. I also am going to add time and key signature there. Do you think that data should be on each track instead?

**Comment by rualark** [comment]
I would suggest to use MIDI type 1 specification.

A type 1 MIDI file should have two or more tracks. The first, by convention, contains song information such as the title, time signature, tempo, etc. The second and following tracks contain a title, musical event data, etc. specific to that track.

Using NOTE ON velocity=0 instead of NOTE OFF seems ok, although you will not be able to send aftertouch information, but this is usually not needed. Using note off is usualy a more robust solution. As far as I remember, some software cannot understand note on velocity=0 events. Size of midi file is usually negligible.

I would recommend to export MIDI Type 1 file as Reaper does:

```
MFile 1 3 960
MTrk
0 Meta TrkName "test"
0 TimeSig 4/4 24 8
0 Tempo 321429
960 Tempo 620690
2880 Tempo 285714
2880 Meta TrkEnd
TrkEnd
MTrk
360 On ch=1 n=58 v=96
1680 Off ch=1 n=58 v=0
2040 On ch=1 n=61 v=96
3120 Off ch=1 n=61 v=0
3120 Meta TrkEnd
TrkEnd
MTrk
240 On ch=1 n=59 v=96
1080 On ch=1 n=52 v=96
1320 Off ch=1 n=59 v=0
2520 Off ch=1 n=52 v=0
2520 Meta TrkEnd
TrkEnd
```

**Comment by paulrosen** [comment]
Let me know what you think of the midi output now. Except for possible bugs, I'm not planning on doing any more changes to the midi file.

**Comment by rualark** [comment]
It looks like this now (I use [this service](http://flashmusicgames.com/midi/mid2txt.php) to conver midi to text):

```
MFile 1 5 480
MTrk
0 Tempo 333333
0 KeySig 7 major
0 TimeSig 4/4 24 8
0 Meta TrkEnd
TrkEnd
MTrk
0 PrCh ch=1 p=0
0 On ch=1 n=60 v=105
1905 Off ch=1 n=60 v=0
1920 On ch=1 n=62 v=105
3825 Off ch=1 n=62 v=0
3825 Meta TrkEnd
TrkEnd
MTrk
0 PrCh ch=1 p=0
0 On ch=2 n=65 v=105
1905 Off ch=2 n=65 v=0
1920 On ch=2 n=64 v=105
3825 Off ch=2 n=64 v=0
3825 Meta TrkEnd
TrkEnd
MTrk
0 PrCh ch=1 p=0
0 Meta TrkEnd
TrkEnd
MTrk
0 PrCh ch=1 p=0
0 Meta TrkEnd
TrkEnd

```

How do I make it export track names?
Why are there gaps between notes (e.g. 1905-1920)? Do you remove these gaps under slurs? It is better not to introduce gaps if it is not required. Here is some information when and how gap is introduced by [different notation software](https://docs.google.com/document/d/1w5pOoVx2ZQjC-BZ4icZs0HejSOazgwthOKOTpR2YLKo/edit?usp=sharing).

P.S. It seems that in this version clicking clef and part name no longer work. Could you please fix that?

**Comment by paulrosen** [comment]
Sorry for not mentioning it but I added another parameter `{ selectAll: true }` to get everything selected. If that is not present, then only the time-related items (notes, rests, and bars) are selectable. I have lots of documentation to do before the next release!

I'm not sure why it isn't exporting your track names. Can you send me the ABC string? It worked for my example.

I added the gaps because it sounded better to me. I'm in the process of improving the sound quality, so I'll look at the link you set to improve it.

I can't remember where I left changing tempo - some work has been done on it, but I'll double check where it is.

Dynamics are on my short list. The link you sent will help because what I was trying to figure out is what volume numbers `pp`, etc. should translate to.

**Comment by rualark** [comment]
Thanks, selectAll works.
ABC string:
```
%%barnumbers 1
M:4/4
K:undefined
L:1/16
V: V0 clef=treble name="Sop."
V: V1 clef=treble name="Alt."
V: V2 clef=treble-8 name="Ten."
V: V3 clef=bass name="Bas."
[V: V0]=C16|=D16|z16|z16|
[V: V1]=F16|=E16|z16|z16|
[V: V2]z16|z16|z16|z16|
[V: V3]z16|z16|z16|z16|

```

If you add gaps to non-slurred notes in exported MIDI file, please make sure that gaps have stable length in PPQ (ideally below or equals to 25 of 480 PPQ) and it does not depend on tempo. This is a common approach in most notation software. On the other hand, this means that duration of gaps in seconds between notes will depend on tempo and thus will lead to different audible results - so this approach will usually be bad for playback in your synth.

For this reason I would separate midi file format (should usually have ppq length-stable gaps for consistent import) and midi, adapted to virtual instruments, that is played in javascript (should usually have duration-stable gaps for best sound).

**Comment by paulrosen** [comment]
Do you have the very latest code from the master branch? I see the track names.

And thanks for that analysis! That is a huge help.

**Comment by rualark** [comment]
Yes, I create MIDI with this call:
`ABCJS.synth.getMidiFile(abc, {midiOutputType: 'encoded'})`

**Comment by paulrosen** [comment]
This is what it looks like to me:
![midi-file-tracks](https://user-images.githubusercontent.com/69818/73889282-52f1ba80-483d-11ea-9bd1-3f38e4e2691e.png)
I'll see if I forgot to check a file in or something.

**Comment by paulrosen** [comment]
Will be fixed in 6.0.0-beta.10
